### ISSUE TITLE
POSIX coap_packet_t defined in .c

### ISSUE THREAD

**Issue by JulianSza** [obgm/libcoap#53]
Is there a reason why coap_packet_t is defined in a header file (coap_io.h) for contiki and lwip but it is defined in source (coap_io.c) for posix? 
Copy-pasting it to the header file does not seem to break the build and enables other apps to peek into the packet. 
Is it safe to move it to the header or is it going to come back and haunt me later?:)

**Comment by obgm** [comment]
coap_packet_t should not be visible to applications, therefore it has been placed in the .c file for platform POSIX, but this would be more difficult for the embedded platforms.

Working on the cleanup of platform-specific code, I found that it might be necessary to place coap_packet_t in a header file for platform POSIX as well. The cleanest solution would be to place this (together with, e.g., the coap_address_t declaration) in a (set of) private header file(s).

That said: Moving it into the header might be necessary but will haunt us later.

**Comment by JulianSza** [comment]
Thanks!:)
Correct me if I am wrong but do you not need to see coap_packet_t if you want to implement a custom network_read in coap_context_t? 

I need to use coap with openssl; I thought the easiest way would be to map ssl_read and write to network_read and write. I know there is a dtls branch with tinydtls but I would like to avoid using it.

**Comment by JulianSza** [comment]
I am struggling to see how is it not needed. 
`network_read` is defined like this:
`ssize_t (*network_read)(coap_endpoint_t *ep, coap_packet_t **packet);` 
`coap_network_read` is allocating and filling in the `coap_packet_t` structure; it requires  `payload`, `length`, `interface` and `src` to work in my case. Whether or not this is a clean solution is a different question:)

TL;DR: libcoap plays nice with OpenSSL if `coap_packet_t` is visible externally.

**Comment by obgm** [comment]
Closing this as resolved because commit 8353417a has moved coap_packet_t into `coap_io.h` for all supported platforms.
